eva/udfs/ndarray/similarity.py

```diff
@@ -0,0 +1,53 @@
+# coding=utf-8
+# Copyright 2018-2022 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import faiss
+import pandas as pd
+
+from eva.udfs.abstract.abstract_udf import AbstractUDF
+
+
+class Similarity(AbstractUDF):
+    def _get_distance(self, numpy_distance):
+        return numpy_distance[0][0]
+
+    def setup(self):
+        pass
+
+    @property
+    def name(self):
+        return "Similarity"
+
+    def forward(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Get similarity score between two feature vectors: 1. feature vector of an opened image;
+        and 2. feature vector from base table.
+        """
+
+        def _similarity(row: pd.Series) -> float:
+            open_feat_np, base_feat_np = (
+                row.iloc[0],
+                row.iloc[1],
+            )
+
+            # Transform to 2D.
```

Comment:
Why do we need to do this transformation?

Reply:
Faiss only accepts a 2D array, so I am doing this transformation.

Reply:
Okay, we should think about enforcing it in the binder. If it requires a lot of effort, we can skip it for now but add a comment.

Reply:
Ok. In this case, if the shape is not 2D (e.g., the feature vector output from the feature extractor), who will take the responsibility to convert it to 2D? The similarity UDF or something else?

Reply:
Is it standard practice to convert the feature vector to 2D for similarity?  I was thinking of reporting it as an error to the user, and they will fix it. But if you think it is standard practise, I'm fine with it.

Reply:
Ok, I will leave a comment for this and we can think about improving this later.

For your approach, if we instead of asking users to take care the transformation, they will probably need another UDF just for reshaping?